Sort patch attachments by filename

Attachments are put in the memory structure depending on the order of promises resolution and `setImpl` calls which leads to flaky tests. For example, https://circleci.com/gh/xodio/xod/1880

```
  1) Load/Save roundtrip xodball -> multifile, multifile -> xodball:

      AssertionError: expected { Object (version, authors, ...) } to deeply equal { Object (version, authors, ...) }
      + expected - actual

           }
           "@/qux": {
             "attachments": [
               {
      -          "content": "# Hello world!\n\n- It's a test file\n- If you see a correct content — it passed!\n- Also it could contain any UTF-8 characters, like this one 😎 or even кириллица!\n\nHave a nice day!\n"
      +          "content": "// Impl loaded fine!\n"
                 "encoding": "utf8"
      -          "filename": "README.md"
      +          "filename": "patch.cpp"
               }
               {
      -          "content": "// Impl loaded fine!\n"
      +          "content": "# Hello world!\n\n- It's a test file\n- If you see a correct content — it passed!\n- Also it could contain any UTF-8 characters, like this one 😎 or even кириллица!\n\nHave a nice day!\n"
                 "encoding": "utf8"
      -          "filename": "patch.cpp"
      +          "filename": "README.md"
               }
               {
                 "content": "iVBORw0KGgoAAAANSUhEUgAAABQAAAAUBAMAAAB/pwA+AAAAG1BMVEXMzMyWlpaxsbGqqqq3t7fFxcWjo6OcnJy+vr5AT8FzAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAAMElEQVQImWNgoBFQVkmBka7i4QxQklmRkUGlAUQyqLCHM4SBSYYkoCqVhiSwDioCAPhiB33L/sGeAAAAAElFTkSuQmCC"
                 "encoding": "base64"
      
      at Context.it (test/loadSave.spec.js:87:12)
      at <anonymous>
```

A solution is to enforce stable sorting by attachment filename. To do it, improve in `patch.js`:

- `setPatchAttachments`
- `attachmentLens`

to guarantee proper sorting.

--- a/packages/xod-fs/test/fixtures/some.xodball
+++ b/packages/xod-fs/test/fixtures/some.xodball
@@ -101,14 +101,14 @@
           "content": "# Hello world!\n\n- It's a test file\n- If you see a correct content — it passed!\n- Also it could contain any UTF-8 characters, like this one 😎 or even кириллица!\n\nHave a nice day!\n"
         },
         {
-          "filename": "patch.cpp",
-          "encoding": "utf8",
-          "content": "// Impl loaded fine!\n"
-        },
-        {
           "filename": "img/20x20.png",
           "encoding": "base64",
           "content": "iVBORw0KGgoAAAANSUhEUgAAABQAAAAUBAMAAAB/pwA+AAAAG1BMVEXMzMyWlpaxsbGqqqq3t7fFxcWjo6OcnJy+vr5AT8FzAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAAMElEQVQImWNgoBFQVkmBka7i4QxQklmRkUGlAUQyqLCHM4SBSYYkoCqVhiSwDioCAPhiB33L/sGeAAAAAElFTkSuQmCC"
+        },
+        {
+          "filename": "patch.cpp",
+          "encoding": "utf8",
+          "content": "// Impl loaded fine!\n"
         }
       ]
     }

--- a/packages/xod-project/src/patch.js
+++ b/packages/xod-project/src/patch.js
@@ -103,12 +103,20 @@ export const setPatchDescription = def(
   R.assoc('description')
 );
 
+const attachmentsLens = R.lens(
+  R.prop('attachments'),
+  R.pipe(
+    R.assoc('attachments'),
+    R.over(R.lensProp('attachments'), R.sortBy(R.prop('filename')))
+  )
+);
+
 /**
  * Returns a Patch with associated attachments list.
  */
 export const setPatchAttachments = def(
   'setPatchAttachments :: [Attachment] -> Patch -> Patch',
-  R.assoc('attachments')
+  R.set(attachmentsLens)
 );
 
 /**
@@ -116,7 +124,7 @@ export const setPatchAttachments = def(
  */
 export const getPatchAttachments = def(
   'getPatchAttachments :: Patch -> [Attachment]',
-  R.prop('attachments')
+  R.view(attachmentsLens)
 );
 
 /**
@@ -135,8 +143,6 @@ export const getImpl = def(
     getPatchAttachments
   )
 );
-
-const attachmentsLens = R.lensProp('attachments');
 
 /**
  * Returns true if patch has a native implementation attached.

